atdpy/test/python-expected/everything.py

```diff
@@ -406,6 +406,58 @@ def to_json_string(self, **kw: Any) -> str:
         return json.dumps(self.to_json(), **kw)
 
 
+@dataclass
+class X2:
+    """Original type: _2"""
+
+    value: Tuple[Kind, Kind, int]
+
+    @classmethod
+    def from_json(cls, x: Any) -> 'X2':
+        return cls((lambda x: (Kind.from_json(x[0]), Kind.from_json(x[1]), _atd_read_int(x[2])) if isinstance(x, list) and len(x) == 3 else _atd_bad_json('array of length 3', x))(x))
+
+    def to_json(self) -> Any:
+        return (lambda x: [(lambda x: x.to_json())(x[0]), (lambda x: x.to_json())(x[1]), _atd_write_int(x[2])] if isinstance(x, tuple) and len(x) == 3 else _atd_bad_python('tuple of length 3', x))(self.value)
+
+    @classmethod
+    def from_json_string(cls, x: str) -> 'X2':
+        return cls.from_json(json.loads(x))
+
+    def to_json_string(self, **kw: Any) -> str:
+        return json.dumps(self.to_json(), **kw)
+
+
+@dataclass
+class X1:
+    """Original type: _1 = { ... }"""
+
+    field_a: int
+    field_b: List[float]
+
+    @classmethod
+    def from_json(cls, x: Any) -> 'X1':
+        if isinstance(x, dict):
+            return cls(
+                field_a=_atd_read_int(x['field_a']) if 'field_a' in x else _atd_missing_json_field('X1', 'field_a'),
+                field_b=_atd_read_list(_atd_read_float)(x['field_b']) if 'field_b' in x else [],
+            )
+        else:
+            _atd_bad_json('X1', x)
+
+    def to_json(self) -> Any:
+        res: Dict[str, Any] = {}
+        res['field_a'] = _atd_write_int(self.field_a)
+        res['field_b'] = _atd_write_list(_atd_write_float)(self.field_b)
+        return res
+
+    @classmethod
+    def from_json_string(cls, x: str) -> 'X1':
+        return cls.from_json(json.loads(x))
+
+    def to_json_string(self, **kw: Any) -> str:
+        return json.dumps(self.to_json(), **kw)
+
+
```

Comment:
These "anonymous" classes `X1` and `X2` aren't great at all for the Python user. We could get rid of `X1` by inlining the tuples. We'd still have anonymous classes for monomorphized records and variants. If we stick with monomorphization, an idea to improve the situation is to generate better names than `X1`, `X2`, etc.

In [our AST definition](https://github.com/returntocorp/semgrep-interfaces/blob/main/ast_generic_v1.atd), we have:
```ocaml
type 'a wrap_ = ('a * tok)
```

It occurs as `string wrap_`, `int wrap_` etc. for which numbered types `_1`, `_2` etc. are created as the result of monomorphization (`_1` becomes Python class `X1`). Instead of numbered type names, we could generate better names like `string_wrap_` and `int_wrap_`.

The nice thing about monomorphization is that it's done once and for all, benefiting all the target languages. Alternatively, if we drop monomorphization, we'd  have to create polymorphic Python classes and involve higher-order functions like we already do for lists and nullables.